src/Gaufrette/Adapter/OpenStack.php

```diff
@@ -0,0 +1,351 @@
+            // @WARNING: This could attempt to load a large amount of data into memory.
+            return (string) $this->getObject($key)->download();
+        } catch (BadResponseError $e) {
+            if (404 === $e->getResponse()->getStatusCode()) {
+                throw new FileNotFound($key);
+            }
+
+            throw StorageFailure::unexpectedFailure('read', ['key' => $key], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write($key, $content)
+    {
+        try {
+            $this->getContainer()->createObject([
+                'name' => $key,
+                'content' => $content,
+            ]);
+        } catch (BadResponseError $e) {
+            throw StorageFailure::unexpectedFailure(
+                'write',
+                ['key' => $key, 'content' => $content],
+                $e
+            );
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists($key)
+    {
+        try {
+            return $this->getContainer()->objectExists($key);
+        } catch (BadResponseError $e) {
+            throw StorageFailure::unexpectedFailure('exists', ['key' => $key], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function keys()
+    {
+        try {
+            return array_map(function (StorageObject $object) {
+                return $object->name;
+            }, iterator_to_array($this->getContainer()->listObjects()));
+        } catch (BadResponseError $e ) {
+            throw StorageFailure::unexpectedFailure('keys', [], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listKeys($prefix = '')
+    {
+        try {
+            return array_filter($this->keys(), function ($key) use ($prefix) {
+                return 0 === strpos($key, $prefix);
+            });
+        } catch (StorageFailure $e) {
+            throw StorageFailure::unexpectedFailure('listKeys', ['prefix' => $prefix], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function mtime($key)
+    {
+        try {
+            return (new \DateTime($this->retrieveObject($key)->lastModified))->format('U');
+        } catch (BadResponseError $e) {
+            if (404 === $e->getResponse()->getStatusCode()) {
+                throw new FileNotFound($key);
+            }
+
+            throw StorageFailure::unexpectedFailure('mtime', ['key' => $key], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete($key)
+    {
+        try {
+            $this->getObject($key)->delete();
+        } catch (BadResponseError $e) {
+            if (404 === $e->getResponse()->getStatusCode()) {
+                throw new FileNotFound($key);
+            }
+
+            throw StorageFailure::unexpectedFailure('delete', ['key' => $key], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function rename($sourceKey, $targetKey)
```

Comment:
You need to check if source file exists too.

Reply:
done ;)